CEDARS hourly output post processing scripts update (New Feature)

lyt5410 opened the pull request:
# Pull Request (PR) Description
**NOTE: SUBMIT A PLAN TO DEER SUPPORT PRIOR TO WORKING ON CODE FOR A MEASURE, FEATURE, OR BUG TO BE CONNECTED WITH OTHERS WORKING ON THE SAME ISSUE OR TO RECEIVE FEEDBACK ON YOUR PROPOSED METHOD**


New Feature Summary

This is the update to 4 existing post-processing python scripts and accompanying data tables in the folder _scripts\data transformation_ to produce hourly loadshape with CEDARS desired format in addition to their original functionality.

Updated the following scripts:
- SFm.py
- MFm.py
- DMo.py
- Com.py

Additional output table(s) that will be created by the update:
- CEDARS_long_ls_SFm.zip
- CEDARS_long_ls_MFm.zip
- CEDARS_long_ls_DMo.zip
- CEDARS_long_ls_Com.zip

Related data tables updates:
- normalizing unit related updates in DEER_EnergyPlus_Modelkit_Measure_list_working.xlsx
- normalizing unit related updates in Normunits.xlsx

These were developed using the following test run cases in the Modelkit system:

For SFm: 
**residential measures\SWBE013-01 Efficient Doors\SWBE013-01 Efficient Doors_SFm_1975**
- climate zones: CZ08, CZ09 
- cohort: SFm&1&rDXHP&Ex&BldgShell, SFm&2&rDXHP&Ex&BldgShell
- cases: RB-BldgShell-Door-Ex, RB-BldgShell-addDoorEff

**residential measures\SWBE013-01 Efficient Doors\SWBE013-01 Efficient Doors_SFm_1985**
- climate zones: CZ10, CZ11
- cohort: SFm&1&rDXHP&Ex&BldgShell, SFm&2&rDXHP&Ex&BldgShell
- cases: RB-BldgShell-Door-Ex, RB-BldgShell-addDoorEff

For MFm:
**residential measures\SWBE013-01 Efficient Doors\SWBE013-01 Efficient Doors_MFm_Ex**
- climate zones: CZ02, CZ03, CZ04
- cohort: MFm&0&rDXGF&Ex&BldgShell, MFm&0&rDXHP&Ex&BldgShell
- cases: RB-BldgShell-Door-Ex, RB-BldgShell-addDoorEff

For DMo:
**residential measures\SWBE013-01 Efficient Doors\SWBE013-01 Efficient Doors_DMo**
- climate zones: CZ02, CZ03
- cohort: DMo&0&rDXGF&Ex&BldgShell, DMo&0&rDXHP&Ex&BldgShell
- cases: RB-BldgShell-Door-Ex, RB-BldgShell-addDoorEff

For Com:
**commercial measures\SWHC024-06 Fan Belt\SWHC024-06 Fan Belt_Ex**
- climate zones: CZ02, CZ03
- cohort: (all available bldgtype cohorts)
- cases: NE-HVAC-FanBelt-SynchVBelt-Msr, NE-HVAC-FanBelt-VBelt-Std


### PR Author
- [x] Make sure the PR branch is up to date with main branch at the time of the PR submission
- [x] Craft a succinct title that effectively encapsulates the essence of the pull request, providing a general overview of the proposed changes.
- [x] Label the PR with at least one of the following: New Measure, Bug, or Feature.
- [x] Provide a concise description of the measure, bug, or feature. Submit one PR per measure.
- [ ] For a new measure, attach a workbook named DEER_EnergyPlus_Modelkit_Measure_list_working.xlsx, containing only rows used for post-processing the measure.
- [x] Add comments in the code when necessary to facilitate the review process.
- [x] Add a comment before the added code, including the author's full name, company, and specifying if it's a bug fix, new measure, or feature.
- [x] For a new feature or bug, demonstrate the impact on energy consumption for selected cases with justification using plots and descriptions.
- [ ] For a new measure, add a summary table showing total energy consumption per simulated case.
### PR Reviewer
- [ ] Conduct a thorough code review.
- [ ] If the branch is behind the main, merge the branch locally to check for potential conflicts.
- [ ] If a bug, locally reproduce it and compare energy consumptions before and after.
- [ ] Explore creative ways to stress-test the code.
- [ ] Locally check the error file and other outputs.

Commit:
initial commit

Commit:
Updated normunits table + DMo script update

Updated DMo CEDARS 8760 output and relevant tables

Commit:
DMo and Com script update

final updates for CEDARS formatted 8760 output

Commit:
SFm script update

updates for CEDARS formatted 8760 output

Commit:
MFm script update

updates for CEDARS formatted 8760 output

Commit:
small updates to Com script

handle if there are no available normalizing units, and tested all Com building types using Econ controls measure

lyt5410 committed:
Merge branch 'sound-data:main' into CEDARS_long_output_update

Commit:
Update Readme on the new CEDARS format

Commit:
Additional update to Readme.md

Commit:
added example input file for Com.py script

Commit:
Merge remote-tracking branch 'upstream/main' into CEDARS_long_output_update

Commit:
updates to DMo.py, normunit fix

Commit:
updates to MFm.py, normunit fix

Commit:
updates to SFm.py, normunit fix

Commit:
updates to Com.py

normunit updates, helper function to extract starting day of week for simulation for CEDARS format

Commit:
updates to 4 scripts + misc. clean ups

Normunit updates (placeholder value for Commercial Cap-tons)
Helper function to extract starting day of week (for Source Year) for simulation for CEDARS format
Rearranged CEDARS code snippet position for better traceability
Updated gitignore for generated outputs

Commit:
Updated scripts for zip output capability, bug fixes

Commit:
update CEDARS output name, update Readme

Commit:
clean up input lines

lyt5410 committed:
Delete scripts/data transformation/CEDARS_long_ls_Com.csv

deleting generated files to the commit

lyt5410 committed:
Delete scripts/data transformation/CEDARS_long_ls_Com_all_bldgtypes_test.csv

deleting generated files to the commit

lyt5410 committed:
Delete scripts/data transformation/CEDARS_long_ls_DMo.csv

deleting generated files to the commit

lyt5410 committed:
Delete scripts/data transformation/CEDARS_long_ls_MFm.csv

deleting generated files to the commit

lyt5410 committed:
Delete scripts/data transformation/CEDARS_long_ls_SFm.csv

deleting generated files to the commit

lyt5410 committed:
Delete scripts/data transformation/test_long_ls.csv

deleting generated files to the commit

lyt5410 committed:
Delete scripts/data transformation/test_long_ls_2.csv

deleting generated files to the commit

lyt5410 committed:
Delete scripts/data transformation/long_ls_test.csv

deleting generated files to the commit

lyt5410 committed:
Merge branch 'main' into CEDARS_long_output_update

Commit:
Merge branch 'main' into pr/133

lyt5410 commented:
Update branch up-to-date with respect to current main, resolved conflicts

nfette commented:
Here are some notes from testing with residential measure SWHC04-08.

- SWHC049-08 DMo: The script finished and created the output files.
- The script takes around 7 minutes to run for DMo building type.

nfette commented:
Here are some notes from testing with commercial measure SWHC046-06.

- After around 15-30 minutes, I walked away. Later, the script failed to finish due to memory at line 530 inside FOR loop from line 501 (see error message below). I was running it on a Windows virtual machine with 16 GB memory. I see a lot of large dataframe variables in memory.
- I switched to a memory-optimized virtual machine with 32 GB memory. The script spends a lot of time running through the FOR loop around line 501. After 135 minutes, the script reached "column 2930 long format loaded" (approx 21 models/minute). After 480 minutes, it reached "col 7600 long format loaded" (approx 16 models/minute) and Python is using as much as 21 GB memory. After around 540 minutes, the script crashed with another memory error at line 629.
- This seems very slow. The script only utilizes 1 CPU core, and is inefficient in applying transformations to data one file at a time, and continuously concatenating to large dataframes. There are 8096 models for this measure. Assuming the script is processing models at a rate of 15-16 models/minute, the script would take over 8-9 hours to run. This is a concern because the script does not save intermediate progress, so any error would lose all progress, and we also tend to run simulations on a virtual machine with many CPU cores, so it is inconvenient or expensive to run a script for so long.

**Memory crash 1 at line 501**
```
#12/22/2025 CEDARS Hourly consumption output reformatting
# use the hourly data before long2wide pivot transform ...
reformatting hourly data for CEDARS loadshape format..
col 0 long format loaded.
col 1 long format loaded.
col 2 long format loaded.
...
col 5417 long format loaded.
col 5418 long format loaded.
col 5419 long format loaded.
---------------------------------------------------------------------------
MemoryError                               Traceback (most recent call last)
e:\CEDARS_long_output_update\scripts\data transformation\Com.py in <module>
     36     hrly_mapped['RunPeriod Start Day'] = col_names[4]
     37 
---> 38     converted_long_df = pd.concat([converted_long_df, hrly_mapped])
     39 
     40     print(f"col {i} long format loaded.")

...

MemoryError: Unable to allocate 362. MiB for an array with shape (1, 47487960) and data type object
```

**Memory crash 2 at line 629**
```
#rearrange / true-up columns ...
---------------------------------------------------------------------------
MemoryError                               Traceback (most recent call last)
e:\CEDARS_long_output_update\scripts\data transformation\Com.py in <module>
     20 #final table fields round-up
     21 #note: UEC and Numunits omitted from draft long table in the final table
---> 22 df_long_final = df_long[['Sector', 'BldgType','BldgVint','BldgHVAC','BldgLoc','Normunit',
     23          'Type', 'Source Year', 'TechGroup', 'TechType','TechID',
     24          'Hour of Year','UEC','UECproportion']] 

...

MemoryError: Unable to allocate 2.64 GiB for an array with shape (5, 70920960) and data type float64
```

sepideh-QEA commented on scripts/data transformation/SFm.py:
This branch removed the `MODE_NEW_VINTAGE` guard that exists on `main`, but it still unconditionally remaps `BldgVint` from climate zone here. If `path_new` is used, New-vintage runs will be rewritten to `1975` or `1985`, which would corrupt both the exported CEDARS loadshapes and the downstream `current_msr_mat` / SQL joins.

On `main`, this remap only happens in the non-new-vintage path:
`if MODE_NEW_VINTAGE: pass`
`else:`
`    sim_annual_f['BldgVint'] = sim_annual_f['BldgLoc'].map(cz_vint_dict)`
`    sim_hourly_final['BldgVint'] = sim_hourly_final['BldgLoc'].map(cz_vint_dict)`

I think this needs to preserve `BldgVint='New'` when `path_new` is used.

lyt5410 replied:
updated SFm.py to made sure it keeps this BldgVint guard.

sepideh-QEA commented on scripts/data transformation/Com.py:
```diff
@@ -468,17 +477,193 @@ def end_use_rearrange(df_in):
+df_numunits = df_normunits[df_normunits['Normunit']==normunit]
+#finalize normalizing unit and num based on measure
+
+if len(df_numunits) == 1:
```
When there is exactly one normunit value, the code sets numunits, but numunit_lookup is never defined. The later try block always references numunit_lookup, so single-value normunits hit the exception path and get exported as Normunit='Each' with Numunits=1. That silently produces incorrect UEC values instead of using the actual normalization from Normunits.xlsx.

sepideh-QEA replied:
also from @Justin-DeBlois: measures with multiple intended normunits (example cited: Len-ft_mt vs Len-ft_lt in the mapping workbook) hit this path, and because `numunit_lookup` isn’t defined the `try/except` silently defaults to `Normunit='Each'` / `Numunits=1`. That produces “valid-looking” output but with incorrect normalization metadata/UEC. Can we fix this by (1) defining `numunit_lookup` for the single-value case and (2) replacing the broad `except` with explicit validation + a hard error when normunit/numunits can’t be determined?

lyt5410 replied:
for Com.py, I removed sections on CEDARS updated that involved normalizing units.

sepideh-QEA commented on scripts/data transformation/MFm.py:
```diff
@@ -340,30 +352,197 @@ def end_use_rearrange(df_in):
 elif (measure_name == 'Wall Insulation') or (measure_name == 'Ceiling Insulation') or (measure_name == 'Windows'):
     numunits = list(numunits_vals[numunits_vals['Msr'] == measure_name]['Value'])[0]
+    print(f'normunit is {normunit}, numunits is {numunits}.')
 elif measure_name == 'PTAC / PTHP':
```
This branch changed numunits_vals to carry BldgLoc instead of CZ, but the PTAC/PTHP branch still indexes numunits_vals['CZ']. On main those names matched; here they do not, so existing PTAC/PTHP post-processing will fail before producing outputs

lyt5410 replied:
fixed this lookup to refer to 'BldgLoc' instead of 'CZ', and moved this section out of CEDARS update portion (to after it is done), since normalizing unit related code are removed from the CEDARS update portion.

sepideh-QEA commented on scripts/data transformation/DMo.py:
```diff
@@ -254,12 +264,16 @@ def end_use_rearrange(df_in):
+    idf_path = hrly_path + "/" + split_meta_cols_eu.iloc[i][0] + "/" + split_meta_cols_eu.iloc[i][1] + "/" + split_meta_cols_eu.iloc[i][2] + "/instance.idf"
+    runperiod_start_day = helper_functions.get_runperiod_start_day(idf_path)
+
     #extract the last column (the total elec hrly profile)
```
On main, these scripts explicitly read Electricity:Facility [J](Hourly). This PR switches to df.iloc[:, -1], which makes the export depend on whatever column happens to be last in instance-var.csv. If the report variable order changes or another hourly series is appended, the CEDARS loadshape and the existing hourly post-processing will silently use the wrong signal.

sepideh-QEA commented:
@nfette Thanks for the testing notes. I think the runtime and memory behavior you saw point to a real scalability issue, not just a VM-specific problem. The current implementation appears to build very large long-format dataframes by repeated concatenation and then materializes another large final dataframe, so the behavior you observed looks plausible for large commercial measures.

sepideh-QEA commented on scripts/data transformation/Com.py:
```diff
@@ -468,17 +477,193 @@ def end_use_rearrange(df_in):
+}
+
+df_long['Sector'] = 'Com' #this is Com script, so Sector = Com
+df_long['Type'] = 'Whole Building'
```
Type is currently hard-coded to Whole Building because the script is exporting whole-building electricity (facility) only. End-use-specific Type would require end-use hourly outputs (lights/HVAC/etc.) plus additional mapping/aggregation logic, which isn’t available in the current post-processing inputs and would expand file size/memory load. Suggest keeping Whole Building for PR #133; we can consider an end-use extension later if/when end-use series are included in the simulation outputs.

Commit:
Update hourly column selection robustness

Commit:
Update hourly data parsing methods for memory efficiency

updated the methods for parsing 365x24 hourly and CEDARS 1-col 8760 from looping pandas dataframe operations to looping data using numpy arrays and dictionaries only, then create dataframe once after.

lyt5410 commented:
> Here are some notes from testing with commercial measure SWHC046-06.
> 
> * After around 15-30 minutes, I walked away. Later, the script failed to finish due to memory at line 530 inside FOR loop from line 501 (see error message below). I was running it on a Windows virtual machine with 16 GB memory. I see a lot of large dataframe variables in memory.
> * I switched to a memory-optimized virtual machine with 32 GB memory. The script spends a lot of time running through the FOR loop around line 501. After 135 minutes, the script reached "column 2930 long format loaded" (approx 21 models/minute). After 480 minutes, it reached "col 7600 long format loaded" (approx 16 models/minute) and Python is using as much as 21 GB memory. After around 540 minutes, the script crashed with another memory error at line 629.
> * This seems very slow. The script only utilizes 1 CPU core, and is inefficient in applying transformations to data one file at a time, and continuously concatenating to large dataframes. There are 8096 models for this measure. Assuming the script is processing models at a rate of 15-16 models/minute, the script would take over 8-9 hours to run. This is a concern because the script does not save intermediate progress, so any error would lose all progress, and we also tend to run simulations on a virtual machine with many CPU cores, so it is inconvenient or expensive to run a script for so long.
> 
> **Memory crash 1 at line 501**
> 
> ```
> #12/22/2025 CEDARS Hourly consumption output reformatting
> # use the hourly data before long2wide pivot transform ...
> reformatting hourly data for CEDARS loadshape format..
> col 0 long format loaded.
> col 1 long format loaded.
> col 2 long format loaded.
> ...
> col 5417 long format loaded.
> col 5418 long format loaded.
> col 5419 long format loaded.
> ---------------------------------------------------------------------------
> MemoryError                               Traceback (most recent call last)
> e:\CEDARS_long_output_update\scripts\data transformation\Com.py in <module>
>      36     hrly_mapped['RunPeriod Start Day'] = col_names[4]
>      37 
> ---> 38     converted_long_df = pd.concat([converted_long_df, hrly_mapped])
>      39 
>      40     print(f"col {i} long format loaded.")
> 
> ...
> 
> MemoryError: Unable to allocate 362. MiB for an array with shape (1, 47487960) and data type object
> ```
> 
> **Memory crash 2 at line 629**
> 
> ```
> #rearrange / true-up columns ...
> ---------------------------------------------------------------------------
> MemoryError                               Traceback (most recent call last)
> e:\CEDARS_long_output_update\scripts\data transformation\Com.py in <module>
>      20 #final table fields round-up
>      21 #note: UEC and Numunits omitted from draft long table in the final table
> ---> 22 df_long_final = df_long[['Sector', 'BldgType','BldgVint','BldgHVAC','BldgLoc','Normunit',
>      23          'Type', 'Source Year', 'TechGroup', 'TechType','TechID',
>      24          'Hour of Year','UEC','UECproportion']] 
> 
> ...
> 
> MemoryError: Unable to allocate 2.64 GiB for an array with shape (5, 70920960) and data type float64
> ```

@nfette I made some revamps to the hourly data parsing methods (both the 365x24 step and the CEDARS long 8760 step) in just the commercial script **Com.py** such that it will no longer loop over multiple pd.DataFrame calls and only loops data over numpy arrays + dictionaries then do 1 dataframe creation after all data has been parsed, I think this should reduce the memory cost. It was a rookie mistake that I have overlooked for scalability - thanks for spending the time to test this out.

Do you mind testing your commercial cohort run again with my updates and let me know how it goes? I assume you still have the simulation results set on your machine.

If this solves the issue, I will apply this to the other 3 scripts. If not, we can work something out.

Commit:
handle trailing spaces issues in col name in instance-var.csv

nfette commented on scripts/data transformation/Com.py:
```diff
@@ -468,17 +506,224 @@ def end_use_rearrange(df_in):
+
+    print(f"col {i} long format loaded.")
+
+#build dataframe once
+final_data = {k: np.concatenate(v) for k, v in long_data.items()}
+converted_long_df = pd.DataFrame(final_data)
```
@lyt5410 Everything up to this point runs much more quickly now, thanks. At line 591, I am running out of memory (32 GB RAM). But there might not be a great way to avoid this. With 128 GB RAM, the script finished without error and took about 25 minutes.

lyt5410 replied:
@sepideh-QEA See my updates and @nfette 's comments on the fix. The main memory issue is solved but with the caveat that a 128 GB machine is required to run a large full Commercial run. Is this an acceptable fix? Or do we want to further develop a work-around for this (such as slicing up the data to output CSVs per building type, using Parquet to store data, etc.) so that any PA/measure developers can do this process with a full commercial run as big as @nfette 's test case here with a 32GB machine.

@nfette can I get a sense of how many permutations/simulations is your test batch run?

In the meantime, I will be fixing the other issues (normunits, etc.).

nfette replied:
I was testing with commercial measure SWHC046-06, which has 8096 model instances.

I think we can work with the proposed fix. During production, we might try to run the script with a small batch of models to make sure there are no errors, then proceed to run with all models.

lyt5410 replied:
Sounds good. I made this update to the Res scripts as well for consistency.

Commit:
Com.py removed UEC related steps for CEDARS loadshape

Commit:
Com.py update: removed Normunit field for CEDARS output

Commit:
SFm.py updates to follow main branch methods

MODE_NEW_VINTAGE guards

Commit:
Added option to output annual consumptions only

Output annual only (1 value per permutation), future implication: for CEDARS loadshape plotly viewers

Commit:
Update PTAC norm unit selection to follow same climate zone label (BldgLoc)

made sure the field lookup is consistent with Normunit.xlsx

lyt5410 commented:
In the process of QC'ing scripts and making sure everything connects, I noticed that the Fan Belt measure folder (SWHC024-06 Fan Belt, the one I was testing for Com script) has some inconsistent labeling on the BldgHVAC component - "rDXHP" instead of "cDXHP" despite being a commercial measure. The scripts will still run but this discrepant descriptor/labelling is carried along the way, which is why I did not note this earlier. Maybe consider adding a check somewhere along the data process pipeline in the future.
<img width="453" height="543" alt="image" src="https://github.com/user-attachments/assets/86f5ea80-89d1-4312-9b7e-657c7aa48bca" />

Commit:
MFm.py rearranged normunit section, CEDARS removed normunit related code

Commit:
fix CEDARS annual output name

Commit:
DMo.py rearranged normunit section, CEDARS removed normunit related code

Commit:
SFm.py rearranged normunit section, CEDARS removed normunit related code

Commit:
clarify comments on CEDARS output

Commit:
DMo.py - Update hourly data parsing methods for memory efficiency

Commit:
cleaned DMo.py some commented code

Commit:
MFm.py - Update hourly data parsing methods for memory efficiency

Commit:
SFm.py - Update hourly data parsing methods for memory efficiency

Commit:
clarified some points on Readme

avoided mentioning exact line number as it can change from script edits

lyt5410 committed:
Merge branch 'main' into CEDARS_long_output_update

Commit:
ensure output exports for plotly viewer

Commit:
Merge branch 'CEDARS_long_output_update' of https://github.com/lyt5410/DEER-Prototypes-EnergyPlus into CEDARS_long_output_update

Commit:
add plotly html generation script for loadshape results

added Readme section, updated requirements.txt

Commit:
Readme heading

Commit:
Readme formatting

Commit:
Readme edits for clarity

Commit:
Update Readme.md

Commit:
display/text edits

sepideh-QEA commented on scripts/data transformation/Com.py:
@lyt5410 this hasn't been resolved. There is no explicit validation right after reading Normunits.xlsx to confirm total_area_m2 exists on the Com sheet. So if a user has an older/different Normunits.xlsx (or the wrong sheet layout), they can still get the same late-stage failure: CEDARS zip created first, then crash here before writing the 3 standard DEER CSVs.

Can you add a clear check like “if total_area_m2 not in df_normunits.columns: raise ValueError(…)” and note that the PR’s updated Normunits.xlsx is required for the commercial path.

Commit:
Update Com.py on normunits validation and area lookup

Normunit validation: added a check to see if measure unit exists in Normunits.xlsx. If not, proceed process with placeholder "Each" and numunit = 1.

meas_area lookup - changed to dictionary lookup as this step only needs to add building-type specific area to the table.

Commit:
Revert "display/text edits"

This reverts commit 6e6951ac37ffa1d0ee9a5ff10c457645cebd1ba2.

Commit:
Revert "Update Readme.md"

This reverts commit 82f16be2fa9b8b8fe17347b60fdc86893affc990.

Commit:
Revert "Readme edits for clarity"

This reverts commit eaeb5f941707655801a155c0de141654c3aadbb8.

Commit:
Revert "Readme formatting"

This reverts commit bf14c1f427284d92a3d6a0e375cc91be8a1f2883.

Commit:
Revert "Readme heading"

This reverts commit c737c5a3145649af8152e021afa93559219cad54.

Commit:
Revert "add plotly html generation script for loadshape results"

This reverts commit 6dc30c296d8dbb18419c75a8f60081f369d30214.

Commit:
Updated Normunits.xlsx

added a separate worksheet Com_area to avoid confusion with normalizing units that have area as the value

Commit:
validation involving Normunits draft

error handling on various scenarios on norm units:
missing normunit in starting table, propoesd normunit doesn't exist in lookup table, wrong datatype normunit

also for scenarios in adding measarea:
file missing, worksheet missing, empty worksheet

Commit:
comments/annotations alongside the NormUnit section

added thoughts for edge cases and future enhancements

Commit:
Updated Readme on the NormUnit stuff
